packages/media-utils/src/components/media-upload-modal/index.tsx

```diff
@@ -0,0 +1,329 @@
+ * @param props.multiple      Whether multiple files can be selected
+ * @param props.value         Currently selected media item(s)
+ * @param props.onSelect      Function called when media is selected
+ * @param props.onClose       Function called when modal is closed
+ * @param props.title         Title for the modal
+ * @param props.isOpen        Whether the modal is open
+ * @param props.isDismissible Whether modal can be dismissed
+ * @param props.modalClass    Additional CSS class for modal
+ * @param props.search        Whether to show search input
+ * @param props.searchLabel   Label for search input
+ * @return JSX element or null
+ */
+export function MediaUploadModal( {
+	allowedTypes = [ 'image' ],
+	multiple = false,
+	value,
+	onSelect,
+	onClose,
+	title = __( 'Select Media' ),
+	isOpen,
+	isDismissible = true,
+	modalClass,
+	search = true,
+	searchLabel = __( 'Search media' ),
+}: MediaUploadModalProps ) {
+	const [ selection, setSelection ] = useState< string[] >( () => {
+		if ( ! value ) {
+			return [];
+		}
+		return Array.isArray( value )
+			? value.map( String )
+			: [ String( value ) ];
+	} );
+
+	// DataViews configuration - allow view updates
+	const [ view, setView ] = useState< View >( () => ( {
+		type: LAYOUT_PICKER_GRID,
+		fields: [],
+		titleField: 'title',
+		mediaField: 'url',
+		search: '',
+		page: 1,
+		perPage: 20,
+		filters: [],
+	} ) );
+
+	// Build query args based on view properties, similar to PostList
+	const queryArgs = useMemo( () => {
+		const filters: Record< string, any > = {};
+
+		view.filters?.forEach( ( filter ) => {
+			// Handle media type filters
+			if ( filter.field === 'media_type' ) {
+				filters.media_type = filter.value;
+			}
+			// Handle author filters
+			if ( filter.field === 'author' ) {
+				filters.author = filter.value;
+			}
+			// Handle date filters
+			if ( filter.field === 'date' ) {
+				filters.after = filter.value?.after;
+				filters.before = filter.value?.before;
+			}
+			// Handle mime type filters
+			if ( filter.field === 'mime_type' ) {
+				filters.mime_type = filter.value;
+			}
+		} );
+
+		// Base media type on allowedTypes if no filter is set
+		if ( ! filters.media_type ) {
+			filters.media_type = allowedTypes.includes( '*' )
+				? undefined
+				: allowedTypes[ 0 ];
+		}
+
+		return {
+			per_page: view.perPage || 20,
+			page: view.page || 1,
+			status: 'inherit',
+			order: view.sort?.direction,
+			orderby: view.sort?.field,
+			search: view.search,
+			...filters,
+		};
+	}, [ view, allowedTypes ] );
+
+	// Fetch all media attachments using WordPress core data with permissions
+	const {
+		records: mediaRecords,
+		isResolving: isLoading,
+		totalItems,
+		totalPages,
+	} = useEntityRecordsWithPermissions( 'postType', 'attachment', queryArgs );
+
+	// Transform the records to the expected format
+	const mediaItems = useMemo( () => {
+		return ( mediaRecords || [] ).map( ( item: any ) =>
+			transformAttachment( item )
```

Comment:
So at the moment the media items passed into `DataViewsPicker` are being transformed? I'm not sure about this choice, I think `DataViewsPicker` should work with something closer to REST API / core data style records. That way it's more likely to be compatible with fields and other shared DataViews functionality. Maybe I'm not understanding the permutations though.

`transformAttachment` could still be used for the selected items passed to the callback to make sure it's compatible with the expected results of the old media modal.

Reply:
Updated.